BQSR refactor for perf improvements

fnothaft opened the pull request:
Large refactor of BQSR targeting #1358. Still a WIP, but pushing for review. TODOs:

- [x] Further documentation pass in BQSR core
- [x] Split classes in BQSR core out to one per file where logical to improve navigability of the codebase
- [x] Complete verification
- [x] See if I can yank out the `QualityScore` and `DecadentRead` classes.

fnothaft commented:
Also, the commits show up out of order here relative to what I get locally, so beware...

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1838/
Test PASSed.

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/models/SnpTable.scala:
```diff
@@ -17,52 +17,111 @@
+                                    idx: Int = 0): Option[Int] = {
+    if (length == 0) {
+      None
+    } else if (rr.start <= sites(offset + idx) && rr.end > sites(offset + idx)) {
```
some inline docs describing what these conditions mean might be useful

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/models/SnpTable.scala:
```diff
@@ -76,20 +135,64 @@ object SnpTable {
 
   /**
-   * Creates a SNP Table from an RDD of RichVariants.
+   * Creates a SNP Table from a VariantsRDD.
```
`VariantsRDD` &rarr; `VariantRDD`

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/DinucCovariate.scala:
```diff
@@ -31,36 +34,62 @@ private[adam] class DinucCovariate extends AbstractCovariate[(Char, Char)] {
       case 'C' => 'G'
       case 'G' => 'C'
-      case 'W' | 'S' | 'Y' | 'R' | 'M' | 'K' | 'B' | 'D' | 'V' | 'H' | 'N' => 'N'
+      case 'N' => 'N'
```
+1

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1841/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1842/
Test PASSed.

devin-petersohn commented on adam-core/src/main/scala/org/bdgenomics/adam/models/SnpTable.scala:
```diff
@@ -17,52 +17,111 @@
   }
 
-  private val unknownContigs = new mutable.HashSet[String]
+  @tailrec private def binarySearch(rr: ReferenceRegion,
```
The next few methods look eerily similar to bdg-utils/IntervalArray methods. Can we not overload the IntervalArray method to use a length and offset? Should SnpTable implement IntervalArray?

fnothaft replied:
Opened #1429 for this.

fnothaft replied:
The TL,DR is that since we're doing a set operation and not a full join, there's a significant performance reason to prefer this more specialized implementation.

devin-petersohn replied:
Thanks!

devin-petersohn commented on adam-core/src/main/scala/org/bdgenomics/adam/models/SnpTable.scala:
```diff
@@ -17,52 +17,111 @@
 /**
  * A table containing all of the SNPs in a known variation dataset.
  *
  * @param table A map between a contig name and a set containing all coordinates
```
I know this is in your TODO, but just to make sure: Docs update.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1847/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1848/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1854/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1855/
Test PASSed.

fnothaft commented:
@heuermh give https://github.com/bigdatagenomics/adam/pull/1423/commits/a40cb7702b21aff7aac0a01daa992a72d3d687fa a look-see.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1856/
Test PASSed.

fnothaft commented:
Yahtzee! Removed both `QualityScore` and `DecadentRead`.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1857/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1860/
Test PASSed.

fnothaft commented:
@heuermh and @devin-petersohn I believe this is good to go, pending a last bit of verification that I will run tomorrow. I think I've addressed all your comments. Can I get a review pass?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1862/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1864/<h2>Build result: ABORTED</span></h2>[...truncated 16 lines...] > /home/jenkins/git2/bin/git rev-parse origin/pr/1423/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 2175619d8748cfc9b007830849ff94c605cf189a # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/1423/merge^{commit} # timeout=10Checking out Revision 2175619d8748cfc9b007830849ff94c605cf189a (origin/pr/1423/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 2175619d8748cfc9b007830849ff94c605cf189aFirst time build. Skipping changelog.Triggering ADAM-prb ? 2.3.0,2.11,1.6.1,centosTriggering ADAM-prb ? 2.6.0,2.10,1.6.1,centosTriggering ADAM-prb ? 2.3.0,2.11,2.0.0,centosTriggering ADAM-prb ? 2.3.0,2.10,2.0.0,centosTriggering ADAM-prb ? 2.6.0,2.11,2.0.0,centosTriggering ADAM-prb ? 2.6.0,2.10,2.0.0,centosTriggering ADAM-prb ? 2.3.0,2.10,1.6.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.6.1,centosADAM-prb ? 2.3.0,2.11,1.6.1,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.10,1.6.1,centos completed with result ABORTEDADAM-prb ? 2.3.0,2.11,2.0.0,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,2.0.0,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.11,2.0.0,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.10,2.0.0,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.6.1,centos completed with result ABORTEDADAM-prb ? 2.6.0,2.11,1.6.1,centos completed with result SUCCESSNotifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1866/<h2>Build result: FAILURE</span></h2>[...truncated 16 lines...] > /home/jenkins/git2/bin/git rev-parse origin/pr/1423/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains cf13794c98bb49cda08051c0079135d711191311 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/1423/merge^{commit} # timeout=10Checking out Revision cf13794c98bb49cda08051c0079135d711191311 (origin/pr/1423/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f cf13794c98bb49cda08051c0079135d711191311First time build. Skipping changelog.Triggering ADAM-prb ? 2.3.0,2.11,1.6.1,centosTriggering ADAM-prb ? 2.6.0,2.10,1.6.1,centosTriggering ADAM-prb ? 2.3.0,2.11,2.0.0,centosTriggering ADAM-prb ? 2.3.0,2.10,2.0.0,centosTriggering ADAM-prb ? 2.6.0,2.11,2.0.0,centosTriggering ADAM-prb ? 2.6.0,2.10,2.0.0,centosTriggering ADAM-prb ? 2.3.0,2.10,1.6.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.6.1,centosADAM-prb ? 2.3.0,2.11,1.6.1,centos completed with result FAILUREADAM-prb ? 2.6.0,2.10,1.6.1,centos completed with result FAILUREADAM-prb ? 2.3.0,2.11,2.0.0,centos completed with result FAILUREADAM-prb ? 2.3.0,2.10,2.0.0,centos completed with result FAILUREADAM-prb ? 2.6.0,2.11,2.0.0,centos completed with result FAILUREADAM-prb ? 2.6.0,2.10,2.0.0,centos completed with result FAILUREADAM-prb ? 2.3.0,2.10,1.6.1,centos completed with result FAILUREADAM-prb ? 2.6.0,2.11,1.6.1,centos completed with result FAILURENotifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

devin-petersohn commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/DinucCovariate.scala:
```diff
@@ -17,60 +17,98 @@
+      //
+      // this is ugly, but as they say,
+      // "At 50, everyone has the face they deserve"
+      require(current == 'A' ||
```
I don't think I understand why we don't allow the fancy bases (S, W, Y, R, etc) here. I know nobody really uses them, but why don't we allow this?

devin-petersohn commented on adam-core/src/test/scala/org/bdgenomics/adam/rdd/read/recalibration/DinucCovariateSuite.scala:
```diff
@@ -0,0 +1,96 @@
+
+  val dc = new DinucCovariate
+
+  test("computing dinucleotide pairs for a single base sequence should return None") {
```
Should probably rename test for refactor. `None` to `('N', 'N')`.

devin-petersohn replied:
`None` to `('N', 'N')`

devin-petersohn commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/DinucCovariate.scala:
```diff
@@ -17,60 +17,98 @@
+        current == 'G' ||
+        current == 'T' ||
+        current == 'N',
+        "Saw invalid base %s. Accepted bases are A,C,G,T,N.".format(current))
```
Why are these the only accepted bases?

fnothaft replied:
We've gone back and forward about this in BQSR over time (e.g., #775). IIRC, Illumina dropped the base resolution codes several years back (unambiguous bases are now just `N`), and we only recalibrate ACTG anyways. From a performance perspective, it's much cheaper to check 5 bases than to check 17.

devin-petersohn commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/DinucCovariate.scala:
```diff
@@ -17,60 +17,98 @@
       case 'C' => 'G'
       case 'G' => 'C'
-      case 'W' | 'S' | 'Y' | 'R' | 'M' | 'K' | 'B' | 'D' | 'V' | 'H' | 'N' => 'N'
+      case _   => 'N'
```
If we really care these odd base cases, we would actually have cases for each of the ambiguous bases. We may want this.

fnothaft replied:
My opinionated view is that we don't care about the resolution codes ;) , and that having the exhaustive match here is preferable from a perf/error catching perspective.

fnothaft commented:
Jenkins, retest this please.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1869/
Test PASSed.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/models/SnpTable.scala:
```diff
@@ -76,20 +134,64 @@ object SnpTable {
+    val (indices, positions) = CollectingSnps.time {
+      val sortedVariants = variants.sort
+        .rdd
+        .cache()
```
I'm inclined to not specify that here, since the SNP table should be much smaller than the reads, and we are caching to preserve the indices.

devin-petersohn replied:
+1

fnothaft commented:
Just pushed a commit that addresses the reviewer comments. I am running the validation overnight and will hopefully have updates in the AM.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1878/
Test PASSed.

heuermh commented on adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala:
```diff
@@ -59,10 +59,10 @@ class TransformArgs extends Args4jBase with ADAMSaveAnyArgs with ParquetArgs {
   var markDuplicates: Boolean = false
   @Args4jOption(required = false, name = "-recalibrate_base_qualities", usage = "Recalibrate the base quality scores (ILLUMINA only)")
   var recalibrateBaseQualities: Boolean = false
+  @Args4jOption(required = false, name = "-min_acceptable_quality", usage = "Minimum acceptable quality for recalibrating a base in a read.")
```
add `, default 5` to usage

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/models/SnpTable.scala:
```diff
@@ -17,52 +17,114 @@
  *
- * @param table A map between a contig name and a set containing all coordinates
- *   where a point variant is known to exist.
+ * @param indices A map mapping contig names to the (first, last) index in the
```
`A map mapping contig names to...` &rarr; `A map of contig names to...`

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/BaseQualityRecalibration.scala:
```diff
@@ -17,115 +17,335 @@
+ *   SNPs to ignore during recalibration table generation.
+ * @param recordGroups The record groups that the reads in this dataset are from.
+ * @param minAcceptableAsciiPhred The minimum acceptable Phred score to attempt
+ *   recalibrating, expressed as an ASCII character with Illumina (33) encodings.
```
add `Defaults to a Phred score of 5.`

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/BaseQualityRecalibration.scala:
```diff
@@ -17,115 +17,335 @@
+    recordGroups: RecordGroupDictionary,
+    minAcceptableQuality: Int,
+    optStorageLevel: Option[StorageLevel]): RDD[AlignmentRecord] = {
+    require(minAcceptableQuality >= 0 && minAcceptableQuality < 93,
```
do we need a similar check above for `minAcceptableAsciiPhred`, or not worth it because it is private API?

fnothaft replied:
Yeah, my thought was to skip the check since it was private.

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/ObservationTable.scala:
```diff
@@ -17,139 +17,35 @@
-
-/**
- * Table containing the empirical frequency of mismatches for each set of covariate values.
+ * @param entries The error covariate -> observed error frequency mapping.
```
-> &rarr; `-&gt;` or `&rarr;`

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/ObservationTable.scala:
```diff
@@ -17,139 +17,35 @@
-  def toCSV: String = {
+  /**
+   * @param recordGroups The record groups that generated the reads in this table.
+   * @return Return this table as CSV compatible with GATK's output.
```
Here and above probably not worth mentioning that the CSV is compatible with GATK's output, because of course they could change it at any time

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/recalibration/RecalibrationTable.scala:
```diff
@@ -0,0 +1,271 @@
+ * In the prior implementation of BQSR, this class was the recalibration table.
+ * However, the downside of this implementation was that the covariate->quality
+ * mapping requires a recursive calculation which winds up running in an inner
+ * loop many times. Given that we expect the number of bases to be orders and
```
```
* However, the downside of that implementation was the covariate &rarr; quality
* mapping required a recursive calculation inside an inner loop many times.
```

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/rich/DecadentRead.scala:
```diff
@@ -1,190 +0,0 @@
-import org.bdgenomics.adam.rich.RichAlignmentRecord._
-import org.bdgenomics.formats.avro.AlignmentRecord
-
-@deprecated("Use RichAlignmentRecord wherever possible in new development.", since = "0.18.0")
```
woot for delete key!

heuermh commented on adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala:
```diff
@@ -59,10 +59,10 @@ class TransformArgs extends Args4jBase with ADAMSaveAnyArgs with ParquetArgs {
+  var minAcceptableQuality: Int = 5
   @Args4jOption(required = false, name = "-stringency", usage = "Stringency level for various checks; can be SILENT, LENIENT, or STRICT. Defaults to LENIENT")
   var stringency: String = "LENIENT"
-  @Args4jOption(required = false, name = "-dump_observations", usage = "Local path to dump BQSR observations to. Outputs CSV format.")
```
If this is gone, do we need any of the `toCSV` stuff below?

fnothaft replied:
The export to CSV functionality is used in the unit tests to validate the recalibrated scores that we assign.

heuermh commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala:
```diff
@@ -629,15 +629,20 @@ case class AlignmentRecordRDD(
-    validationStringency: ValidationStringency = ValidationStringency.LENIENT): AlignmentRecordRDD = BQSRInDriver.time {
-    replaceRdd(BaseQualityRecalibration(rdd, knownSnps, observationDumpFile, validationStringency))
+    minAcceptableQuality: Int = 5,
+    optStorageLevel: Option[StorageLevel] = Some(StorageLevel.MEMORY_ONLY)): AlignmentRecordRDD = BQSRInDriver.time {
```
in #1427 should I have used `optStorageLevel` instead of `storageLevel`?

fnothaft replied:
Oh, probably yeah, but not a huge deal. We should clean that up though.

fnothaft commented:
@heuermh squashed down and addressed review comments.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1886/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1888/
Test PASSed.

heuermh commented:
Thank you, @fnothaft!
